### DISCUSSION TITLE
Resin Mixer Tool

### DISCUSSION THREAD

**Discussion by eefweenink** [sn4k3/UVtools#839] [Ideas]
Today I saw this interesting video https://www.youtube.com/watch?v=OKvJ7RY_S7A about using the buildplate as a "mixer" for the resin in the vat. 
Simple resume: A non-existing object is printed; while the layerheight = 0.001 mm; no (or as low as possible) exposuretime; the liftheight is set to about 20-30 mm (just above the resin); the wait before cure is set to 600 - 1200 or 1800 seconds (so 10-20-or 30 minutes). The effect is that there is NO exposure WHILE every 10-20-30 minutes the plate goes down to the bottom of the vat and is pulled out again. 
Takes about a week to finish this print, so the resin is nicely in motion and a new print can be started anytime without stirring and mixing. 

Simular: A shorter program to stir the resin during 10-15 minutes continuously (no waiting time) to stir up resin what has been standing/resting and has to to be made ready for printing. 

Would be nice to have this as a tool in UVTools. Maybe with 2 types, quick-short and continuously
setting the height of the resin; and the interval in minutes. 

Thanks & regards, Eef Weenink

**Comment by sn4k3** [comment]
You can produce a script to do that with UVtools.
Most printers will not be compatible with stir and print object in same file, however if you create a dummy file just to stir is possible to follow your idea in most of the printers. Just slice a file with desired height and parameters, go to UVtools and set all layers to black (No pixels) with the least exposure time you can go with.

**Reply by eefweenink** [reply]
Correct! Seems it is not hard to create this. 
My suggestion is to add this as a tool to UVTools. so it will be more easy and possible for (all/more) users.

**Reply by sn4k3** [reply]
I can give a look into it

**Reply by sn4k3** [reply]
![image](https://github.com/sn4k3/UVtools/assets/113281/2bdf4ec1-4abe-426f-a688-04497407be55)

Are this options ok? Or you have more that you like to add?

**Comment by eefweenink** [comment]
Looks good to me.
In the video there was a difference between fast en short before printing and slow (1x per 10-30 minutes) for days or a week.
What you show seems to have both options (just setting the waiting time makes the difference. )

regard, Eef

Van: Tiago Conceição ***@***.***>
Datum: maandag, 26 februari 2024 om 18:46
Aan: sn4k3/UVtools ***@***.***>
CC: eefweenink ***@***.***>, Author ***@***.***>
Onderwerp: Re: [sn4k3/UVtools] Resin Mixer Tool (Discussion #839)

image.png (view on web)<https://github.com/sn4k3/UVtools/assets/113281/2bdf4ec1-4abe-426f-a688-04497407be55>

Are this options ok? Or you have more that you like to add?

**Reply by sn4k3** [reply]
You can now test the new tool

**Comment by eefweenink** [comment]
Dear Tiago,


Thank you for letting me test this. I worked on it a couple of hours. These are my results:

Test of: Stir Resin
UVTools

  *   Explanation:
     *   Add: To create a stirring file, first open a file specific to the printer. This will include the correct printer options into the file.
     *   Add: Options:
1. Add the stir to the beginning of the file to print
2. Stir as independent action. (Save file to be re-run again)
     *   “Are you sure you want to modify the file to stir the resin in the VAT using the method StirOnly at print start?”
change to:
Are you sure you want to modify/replace the file to stir the resin in the VAT using the method StirOnly at print start?
     *   After saving it is visible how it works:
First - exposure to 0,0001 layer to 0,0001 and 1 pixel outside recommend area.
  *   Suggestions
        *   Set filename to same as Stir Profile Name (I had to use “rename” to change the name before save)
        *   Stir Resin, unable to run
is shown when there is an existing Stir Profile loaded.
Maybe better to add checkbox with option:
O Delete loaded stir profile? Y/N
        *   What happens if user applies suggested changes? Because layer height and exposure time are outside the printer ranges.
        *   When adding to existing file. Remove the 1 pixel. Because if user changes model position the pixel will influence the printed area. Correct?
  *   All options: Lift height speed retract wait times were greyed out.
So no possibility to change something there.
(Machine is UVTools SLS1S SPEED. I did not load a file for Halot One (my machine) first.
NEW START
  *   I started with old file for Halo One, and options were available.
  *   There is no option to EDIT the profile.  I have to remove all but 1 layers first. Start a new profile and overwrite existing file.
Workaround:  Reopen an old stirring file (now the message Unable to run is not showed). Choose a profile that will NOT print existing file, Save changes to NEW profile, delete old profile.
  *   If user opens a file WITH a stirring action at the start, is is possible to add a second stirring profile etcetera.
Pro: You  may create a complex profile this way: example: Stir first short and heavily, then proceed with 1 stir per 10 minutes or so.
Con: User may accidentally add more stirring actions without knowing.
Maybe better to have a check if there is a stirring action already in the file? AND/OR when saving add “stirred” to filename.
  *   With the Halo One I get this message:
Your printer and/or file format is not compatible with the method StirAndPrint. Please select another method. (when wanting the stir added to beginning of file).
QUESTION: What makes printer compatible?

Testing on the Halot One,
  *   Ai, it did not accept the files I made. Got errormessage (see image) [Afbeelding met beeldscherm, tekst, multimedia, Weergave-apparaat  Automatisch gegenereerde beschrijving]
  *   I changed exposuretime. Layer height, wait time etc. NO SUCCESS.
  *   So started over with a file in Prusaslicer. I placed a small dot on the plate, Then added a negative part all over it, so as result there is nothing on the plate (otherwise Prusaslicer did not accept an empty plate to slice).
  *   Exported as SLS1 to UVtools (file is attached) . In UVtools it is showed as all emptylayers.
  *   In UVTools I made several files, with different parameters. I tested to be sure, Halot One would accept it.
  *   Then I used the Stir Resin tool . I had to set exposure time to mimimum of 1 second.
I can change height, waittime, etc. (I did not test completely, but I don’t think It will accept other speed for lift and rectract then standard.
So with a waittime of 20 sec + liftheight of 45 mm it takes about 1 minute + 30 seconds to do one layer. Enough room to make a long term stirrer or a fast and short stir process.
  *   The last version what workd is also attached.

[TestStir.zip](https://github.com/sn4k3/UVtools/files/14465068/TestStir.zip)



Regards, Eef

**Reply by sn4k3** [reply]
**When you use email to reply on github, all attachments and images are not loaded. Always use web to reply with attachments.**

>  Set filename to same as Stir Profile Name (I had to use “rename” to change the name before save)

Stir is just like another operation, it have no access to rename file. You must use File -> Rename (Ctrl + F2) or File - Save As.
Also such behavior may raise other problems and not be the prefered way for other users. 

> Maybe better to add checkbox with option:
> O Delete loaded stir profile? Y/N

I can allow the StirOnly to always run as it replace the file.
For StirAndPrint is a little more tricky but I will try to do that.

> What happens if user applies suggested changes? Because layer height and exposure time are outside the printer ranges.

It will change the wait times, if you use stironly theres no reason to apply suggestions. For StirAndPrint you should apply before run the tool. This is only valid if you set any wait time to stir tool, otherwise it won't matter.

> When adding to existing file. Remove the 1 pixel. Because if user changes model position the pixel will influence the printed area. 

The pixel must be there for UVtools not complain about empty layer and auto remove it, also some printers must have data on layer. If you move model that pixel will move accordingly, it's not a problem and won't influence new bounds. But if you really want to remove it I can place the option to generate the pixel or not. Still I advice against empty layers.

>  All options: Lift height speed retract wait times were greyed out.

When any option is grey out it means not supported on current setup.

> (Machine is UVTools SLS1S SPEED. I did not load a file for Halot One (my machine) first.

Those machines are potatoes on settings, can't adjust anything (layer based) and one of the worse printers for this kind of tools.

> QUESTION: What makes printer compatible?

The ability to follow layer data with at least PositionZ avaliable

> is also attached.

You have to reattach all in web github

**Reply by eefweenink** [reply]
I attached two files in zip into my message.

**Comment by eefweenink** [comment]
> Those machines are potatoes on settings, can't adjust anything (layer based) and one of the worse printers for this kind of tools.

OK, so I have to live with that ☺
And explains why I cannot attach the stiraction to a file to be printed.

Maybe the files I have attached, help finding why the file (made with only the tool) did not work, and the file I made using an already by the printer accepted file was OK.

Regards, Eef

**Reply by sn4k3** [reply]
From what I understood, **teststir.cxdlp** won't print but **teststir.sl1** after conversion will?

**Comment by eefweenink** [comment]
Wrong file: 
[teststiremptylayer.sl1.zip](https://github.com/sn4k3/UVtools/files/14465808/teststiremptylayer.sl1.zip)
This file was made with Prusaslicer. Imported in UVtools. Exported as cxdlp and printed OK
Then I used that file to experiment with StirResin and  teststir.cxdlp (as I remember) works fine. 
I am sure that setting exposuretimes below 1 second gave an error.

**Reply by sn4k3** [reply]
Maybe, 0.01s is too short to be a valid exposure still we never know what we can expect from this firmwares.
I will allow to set exposure time on next patch. But you can try 0.1s?

**Reply by eefweenink** [reply]
I did try several times. 0.99 and lower gives error. 1 sec is OK (for this printer).

**Reply by sn4k3** [reply]
Have a look in the new version

**Comment by eefweenink** [comment]
Have tested them on attached files: 
[3 stir files for Halot One.zip](https://github.com/sn4k3/UVtools/files/14471880/3.stir.files.for.Halot.One.zip)
In all files: waittime = 20secs: NO output pixel (empty layer): 10 layers 
1. stirCL-60_0.01_20sec.cxdlp was not accepted by the machine (probably the 0.01 exposuretime)
2. stirCL-60_1_1sec.cxdlp run OK. changed exposure time to 1 sec(because of height 45 mm up and down + 20secs waittime 1 cycle takes about 1:16 min:sec
3. stirCL-60_1_1sec_25mm_20wait.cxdlp (changed exposuretime to 1 sec, bottom exposure time changed to 1 sec (was 20) and height changed to 25 mm waittime was 20 secs (run OK. 1 cycle took about 1:01 min:sec)
NB: All expected printtimes (calculated by UVtools and Halot One) were incorrected (guess the long up and down is not calculated exact).

**Reply by sn4k3** [reply]
Ok, so you found the minimum for that machine.
Keep that in mind when creating the tests.
As it is exposed for 1s it is worth to disable the dummy pixel.

**Reply by eefweenink** [reply]
Yep, is het possible to get/save these parameters in the printerprofile? So the tool StirResion can use that?

**Reply by sn4k3** [reply]
> Yep, is het possible to get/save these parameters in the printerprofile? So the tool StirResion can use that? 

Yes, you can add a profile and set that as default.
